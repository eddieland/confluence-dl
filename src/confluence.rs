//! Confluence API client for fetching pages and content.
//!
//! This module provides a client for interacting with the Confluence REST API,
//! including authentication, page fetching, and content retrieval.

use std::time::Duration;

use anyhow::{Context, Result, anyhow};
use base64::Engine as _;
use base64::engine::general_purpose::STANDARD as BASE64;
use serde::{Deserialize, Serialize};
use url::Url;

/// Trait for Confluence API operations (enables testing with fake
/// implementations)
pub trait ConfluenceApi {
  /// Fetch a page by ID
  fn get_page(&self, page_id: &str) -> Result<Page>;

  /// Get attachments for a page
  fn get_attachments(&self, page_id: &str) -> Result<Vec<Attachment>>;

  /// Download an attachment by URL to a file
  fn download_attachment(&self, url: &str, output_path: &std::path::Path) -> Result<()>;

  /// Test authentication
  #[allow(dead_code)]
  fn test_auth(&self) -> Result<()>;
}

/// Confluence API client
pub struct ConfluenceClient {
  base_url: String,
  username: String,
  token: String,
  client: reqwest::blocking::Client,
}

/// Confluence page metadata and content
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Page {
  pub id: String,
  pub title: String,
  #[serde(rename = "type")]
  pub page_type: String,
  pub status: String,
  pub body: Option<PageBody>,
  pub space: Option<PageSpace>,
  #[serde(rename = "_links")]
  pub links: Option<PageLinks>,
}

/// Page body content in various formats
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PageBody {
  pub storage: Option<StorageFormat>,
  pub view: Option<ViewFormat>,
}

/// Storage format (Confluence's internal format)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StorageFormat {
  pub value: String,
  pub representation: String,
}

/// View format (rendered HTML)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ViewFormat {
  pub value: String,
  pub representation: String,
}

/// Space information
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PageSpace {
  pub key: String,
  pub name: String,
  #[serde(rename = "type")]
  pub space_type: String,
}

/// Page links
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PageLinks {
  #[serde(rename = "webui")]
  pub web_ui: Option<String>,
  #[serde(rename = "self")]
  pub self_link: Option<String>,
}

/// Attachment metadata
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Attachment {
  pub id: String,
  pub title: String,
  #[serde(rename = "type")]
  pub attachment_type: String,
  #[serde(rename = "mediaType")]
  pub media_type: Option<String>,
  #[serde(rename = "fileSize")]
  pub file_size: Option<u64>,
  #[serde(rename = "_links")]
  pub links: Option<AttachmentLinks>,
}

/// Attachment links
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AttachmentLinks {
  pub download: Option<String>,
}

/// Attachments response wrapper
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AttachmentsResponse {
  pub results: Vec<Attachment>,
}

/// Information extracted from a Confluence URL
#[derive(Debug, Clone)]
pub struct UrlInfo {
  pub base_url: String,
  pub page_id: String,
  pub space_key: Option<String>,
}

impl ConfluenceClient {
  /// Create a new Confluence client
  ///
  /// # Arguments
  /// * `base_url` - The base URL of the Confluence instance (e.g., https://example.atlassian.net)
  /// * `username` - The user's email address
  /// * `token` - The API token
  /// * `timeout_secs` - Request timeout in seconds
  pub fn new(
    base_url: impl Into<String>,
    username: impl Into<String>,
    token: impl Into<String>,
    timeout_secs: u64,
  ) -> Result<Self> {
    let base_url = base_url.into();
    let username = username.into();
    let token = token.into();

    // Normalize base URL (remove trailing slash)
    let base_url = base_url.trim_end_matches('/').to_string();

    // Create HTTP client with timeout
    let client = reqwest::blocking::Client::builder()
      .timeout(Duration::from_secs(timeout_secs))
      .user_agent(format!(
        "confluence-dl/{} ({})",
        env!("CARGO_PKG_VERSION"),
        env!("TARGET")
      ))
      .build()
      .context("Failed to create HTTP client")?;

    Ok(Self {
      base_url,
      username,
      token,
      client,
    })
  }

  /// Get the authorization header value (Basic auth)
  fn auth_header(&self) -> String {
    let credentials = format!("{}:{}", self.username, self.token);
    format!("Basic {}", BASE64.encode(credentials.as_bytes()))
  }
}

impl ConfluenceApi for ConfluenceClient {
  fn get_page(&self, page_id: &str) -> Result<Page> {
    let url = format!(
      "{}/wiki/rest/api/content/{}?expand=body.storage,body.view,space",
      self.base_url, page_id
    );

    let response = self
      .client
      .get(&url)
      .header("Authorization", self.auth_header())
      .header("Accept", "application/json")
      .send()
      .context("Failed to send request to Confluence API")?;

    if !response.status().is_success() {
      let status = response.status();
      let error_text = response.text().unwrap_or_else(|_| String::from("(no error details)"));
      return Err(anyhow!("Confluence API returned error {status}: {error_text}"));
    }

    let page: Page = response
      .json()
      .context("Failed to parse page response from Confluence API")?;

    Ok(page)
  }

  fn get_attachments(&self, page_id: &str) -> Result<Vec<Attachment>> {
    let url = format!("{}/wiki/rest/api/content/{}/child/attachment", self.base_url, page_id);

    let response = self
      .client
      .get(&url)
      .header("Authorization", self.auth_header())
      .header("Accept", "application/json")
      .send()
      .context("Failed to fetch attachments from Confluence API")?;

    if !response.status().is_success() {
      let status = response.status();
      let error_text = response.text().unwrap_or_else(|_| String::from("(no error details)"));
      return Err(anyhow!("Confluence API returned error {status}: {error_text}"));
    }

    let attachments: AttachmentsResponse = response
      .json()
      .context("Failed to parse attachments response from Confluence API")?;

    Ok(attachments.results)
  }

  fn download_attachment(&self, url: &str, output_path: &std::path::Path) -> Result<()> {
    // Build full URL if it's a relative path
    let full_url = if url.starts_with("http://") || url.starts_with("https://") {
      url.to_string()
    } else {
      format!("{}{}", self.base_url, url)
    };

    let response = self
      .client
      .get(&full_url)
      .header("Authorization", self.auth_header())
      .send()
      .context("Failed to download attachment")?;

    if !response.status().is_success() {
      let status = response.status();
      return Err(anyhow!("Failed to download attachment: {status}"));
    }

    // Create parent directory if it doesn't exist
    if let Some(parent) = output_path.parent() {
      std::fs::create_dir_all(parent).context("Failed to create output directory for attachment")?;
    }

    // Write response bytes to file
    let bytes = response.bytes().context("Failed to read attachment bytes")?;
    std::fs::write(output_path, bytes).context("Failed to write attachment to file")?;

    Ok(())
  }

  fn test_auth(&self) -> Result<()> {
    let url = format!("{}/wiki/rest/api/user/current", self.base_url);

    let response = self
      .client
      .get(&url)
      .header("Authorization", self.auth_header())
      .header("Accept", "application/json")
      .send()
      .context("Failed to send authentication test request")?;

    if !response.status().is_success() {
      let status = response.status();
      return Err(anyhow!("Authentication failed with status: {status}"));
    }

    Ok(())
  }
}

/// Parse a Confluence URL to extract page ID and other information
///
/// Supports various Confluence URL formats:
/// - https://example.atlassian.net/wiki/spaces/SPACE/pages/123456/Page+Title
/// - https://example.atlassian.net/wiki/pages/123456
/// - https://example.atlassian.net/pages/123456
pub fn parse_confluence_url(url: &str) -> Result<UrlInfo> {
  let parsed = Url::parse(url).context("Invalid URL format")?;

  // Extract base URL (scheme + host)
  let base_url = format!(
    "{}://{}",
    parsed.scheme(),
    parsed.host_str().context("URL missing host")?
  );

  // Extract path segments
  let path = parsed.path();
  let segments: Vec<&str> = path.split('/').filter(|s| !s.is_empty()).collect();

  // Look for "pages" segment followed by a numeric ID
  let page_id_pos = segments
    .iter()
    .position(|&s| s == "pages")
    .context("URL does not contain 'pages' segment")?;

  if page_id_pos + 1 >= segments.len() {
    return Err(anyhow!("URL does not contain page ID after 'pages' segment"));
  }

  let page_id = segments[page_id_pos + 1];

  // Verify page ID is numeric
  if !page_id.chars().all(|c| c.is_ascii_digit()) {
    return Err(anyhow!("Page ID is not numeric: {page_id}"));
  }

  // Try to extract space key (appears between "spaces" and "pages")
  let space_key = segments.iter().position(|&s| s == "spaces").and_then(|pos| {
    if pos + 1 < segments.len() && pos + 1 < page_id_pos {
      Some(segments[pos + 1].to_string())
    } else {
      None
    }
  });

  Ok(UrlInfo {
    base_url,
    page_id: page_id.to_string(),
    space_key,
  })
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn test_parse_confluence_url_with_space() {
    let url = "https://eddieland.atlassian.net/wiki/spaces/~6320c26429083bbe8cc369b0/pages/229483/Getting+started+in+Confluence+from+Jira";
    let info = parse_confluence_url(url).unwrap();

    assert_eq!(info.base_url, "https://eddieland.atlassian.net");
    assert_eq!(info.page_id, "229483");
    assert_eq!(info.space_key, Some("~6320c26429083bbe8cc369b0".to_string()));
  }

  #[test]
  fn test_parse_confluence_url_without_space() {
    let url = "https://example.atlassian.net/wiki/pages/123456";
    let info = parse_confluence_url(url).unwrap();

    assert_eq!(info.base_url, "https://example.atlassian.net");
    assert_eq!(info.page_id, "123456");
    assert_eq!(info.space_key, None);
  }

  #[test]
  fn test_parse_confluence_url_invalid() {
    let url = "https://example.com/not-a-confluence-url";
    assert!(parse_confluence_url(url).is_err());
  }

  #[test]
  fn test_parse_confluence_url_non_numeric_id() {
    let url = "https://example.atlassian.net/wiki/pages/notanumber";
    assert!(parse_confluence_url(url).is_err());
  }

  #[test]
  fn test_confluence_client_new() {
    let client = ConfluenceClient::new("https://example.atlassian.net", "user@example.com", "test-token", 30);
    assert!(client.is_ok());
    let client = client.unwrap();
    assert_eq!(client.base_url, "https://example.atlassian.net");
    assert_eq!(client.username, "user@example.com");
    assert_eq!(client.token, "test-token");
  }

  #[test]
  fn test_confluence_client_new_removes_trailing_slash() {
    let client = ConfluenceClient::new("https://example.atlassian.net/", "user@example.com", "test-token", 30).unwrap();
    assert_eq!(client.base_url, "https://example.atlassian.net");
  }

  #[test]
  fn test_auth_header_format() {
    let client = ConfluenceClient::new("https://example.atlassian.net", "user@example.com", "test-token", 30).unwrap();

    let auth_header = client.auth_header();
    assert!(auth_header.starts_with("Basic "));

    // Decode and verify the Base64 encoded credentials
    let encoded = auth_header.strip_prefix("Basic ").unwrap();
    let decoded = BASE64.decode(encoded.as_bytes()).unwrap();
    let decoded_str = String::from_utf8(decoded).unwrap();
    assert_eq!(decoded_str, "user@example.com:test-token");
  }

  #[test]
  fn test_parse_confluence_url_missing_pages_segment() {
    let url = "https://example.atlassian.net/wiki/spaces/SPACE/123456";
    assert!(parse_confluence_url(url).is_err());
  }

  #[test]
  fn test_parse_confluence_url_pages_at_end() {
    let url = "https://example.atlassian.net/wiki/pages";
    let result = parse_confluence_url(url);
    assert!(result.is_err());
    assert!(result.unwrap_err().to_string().contains("does not contain page ID"));
  }

  #[test]
  fn test_parse_confluence_url_invalid_scheme() {
    let url = "not-a-url";
    assert!(parse_confluence_url(url).is_err());
  }

  #[test]
  fn test_parse_confluence_url_no_host() {
    let url = "file:///wiki/pages/123";
    let result = parse_confluence_url(url);
    // This should error because file:// scheme doesn't have a host
    assert!(result.is_err());
  }
}
