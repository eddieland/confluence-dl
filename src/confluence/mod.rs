//! Confluence module providing API abstractions, the HTTP client, data models,
//! URL parsing helpers, and higher-level traversal utilities.

pub mod api;
pub mod client;
pub mod models;
pub mod tree;
pub mod url;

pub use api::ConfluenceApi;
pub use client::ConfluenceClient;
#[allow(unused_imports)]
pub use models::{
  Attachment, AttachmentLinks, AttachmentsResponse, ChildPagesResponse, Page, PageBody, PageLinks, PageSpace,
  StorageFormat, UserInfo, ViewFormat,
};
pub use tree::{PageTree, get_page_tree};
pub use url::{UrlInfo, parse_confluence_url};
