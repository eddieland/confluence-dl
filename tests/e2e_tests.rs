//! End-to-end tests using the fake Confluence client
//!
//! These tests demonstrate complete workflows using the fake client,
//! including page fetching, error handling, and data validation.

mod common;

use common::fake_confluence::FakeConfluenceClient;
use common::fixtures;
use confluence_dl::confluence::ConfluenceApi;
use insta::assert_snapshot;

#[test]
fn test_fetch_basic_page() {
  let client = FakeConfluenceClient::with_sample_pages();

  let page = client.get_page("123456").unwrap();

  assert_eq!(page.id, "123456");
  assert_eq!(page.title, "Getting Started Guide");
  assert_eq!(page.page_type, "page");
  assert_eq!(page.status, "current");

  // Verify body content
  assert!(page.body.is_some());
  let body = page.body.unwrap();
  assert!(body.storage.is_some());

  let storage = body.storage.unwrap();
  assert!(storage.value.contains("Getting Started"));
  assert!(storage.value.contains("Welcome to our documentation"));
}

#[test]
fn test_fetch_complex_page_with_code() {
  let client = FakeConfluenceClient::with_sample_pages();

  let page = client.get_page("789012").unwrap();

  assert_eq!(page.title, "API Documentation");

  let body = page.body.unwrap();
  let storage = body.storage.unwrap();

  // Verify complex content is preserved
  assert!(storage.value.contains("API Documentation"));
  assert!(storage.value.contains("ac:structured-macro"));
  assert!(storage.value.contains("python"));
  assert!(storage.value.contains("import requests"));
}

#[test]
fn test_fetch_page_with_internal_links() {
  let client = FakeConfluenceClient::with_sample_pages();

  let page = client.get_page("345678").unwrap();

  assert_eq!(page.title, "Installation Guide");

  let body = page.body.unwrap();
  let storage = body.storage.unwrap();

  // Verify internal links are present
  assert!(storage.value.contains("ac:link"));
  assert!(storage.value.contains("ri:page"));
  assert!(storage.value.contains("Getting Started Guide"));
  assert!(storage.value.contains("API Documentation"));
}

#[test]
fn test_fetch_personal_space_page() {
  let client = FakeConfluenceClient::with_sample_pages();

  let page = client.get_page("229483").unwrap();

  assert_eq!(page.title, "Getting started in Confluence from Jira");

  // Verify space information
  let space = page.space.unwrap();
  assert_eq!(space.space_type, "personal");
  assert_eq!(space.key, "~6320c26429083bbe8cc369b0");
  assert_eq!(space.name, "Edward Jones");
}

#[test]
fn test_fetch_page_with_images() {
  let client = FakeConfluenceClient::with_sample_pages();

  let page = client.get_page("456789").unwrap();

  assert_eq!(page.title, "Architecture Diagram");

  let body = page.body.unwrap();
  let storage = body.storage.unwrap();

  // Verify image attachments are present
  assert!(storage.value.contains("ac:image"));
  assert!(storage.value.contains("ri:attachment"));
  assert!(storage.value.contains("architecture.png"));
}

#[test]
fn test_fetch_nonexistent_page() {
  let client = FakeConfluenceClient::with_sample_pages();

  let result = client.get_page("999999");

  assert!(result.is_err());
  let err = result.unwrap_err();
  assert!(err.to_string().contains("No content found"));
  assert!(err.to_string().contains("999999"));
}

#[test]
fn test_authentication_success() {
  let client = FakeConfluenceClient::with_sample_pages();

  let result = client.test_auth();
  assert!(result.is_ok());

  let user_info = result.unwrap();
  assert_eq!(user_info.account_id, "test-account-id");
  assert_eq!(user_info.email, Some("test@example.com".to_string()));
  assert_eq!(user_info.display_name, "Test User");
  assert_eq!(user_info.public_name, Some("Test User".to_string()));
}

#[test]
fn test_authentication_failure() {
  let mut client = FakeConfluenceClient::with_sample_pages();
  client.set_auth_success(false);

  let result = client.test_auth();
  assert!(result.is_err());

  let err = result.unwrap_err();
  assert!(err.to_string().contains("Authentication failed"));
  assert!(err.to_string().contains("401"));
}

#[test]
fn test_custom_page_workflow() {
  let mut client = FakeConfluenceClient::new();

  // Start with empty client
  assert!(client.get_page("custom123").is_err());

  // Add a custom page
  client.add_page_from_json("custom123", fixtures::sample_page_response());

  // Now it should be fetchable
  let page = client.get_page("custom123").unwrap();
  assert_eq!(page.title, "Getting Started Guide");
}

#[test]
fn test_multiple_pages_workflow() {
  let client = FakeConfluenceClient::with_sample_pages();

  // Fetch multiple pages in sequence
  let pages = vec!["123456", "789012", "345678"];

  for page_id in pages {
    let page = client.get_page(page_id).unwrap();
    assert_eq!(page.id, page_id);
    assert!(page.body.is_some());
  }
}

#[test]
fn test_page_links_metadata() {
  let client = FakeConfluenceClient::with_sample_pages();

  let page = client.get_page("123456").unwrap();

  // Verify links metadata
  assert!(page.links.is_some());
  let links = page.links.unwrap();

  assert!(links.web_ui.is_some());
  assert!(links.self_link.is_some());

  let web_ui = links.web_ui.unwrap();
  assert!(web_ui.contains("/wiki/spaces/DOCS/pages/123456"));
}

#[test]
fn test_page_space_metadata() {
  let client = FakeConfluenceClient::with_sample_pages();

  let page = client.get_page("789012").unwrap();

  let space = page.space.unwrap();
  assert_eq!(space.key, "DEV");
  assert_eq!(space.name, "Developer Portal");
  assert_eq!(space.space_type, "global");
}

#[test]
fn test_storage_format_representation() {
  let client = FakeConfluenceClient::with_sample_pages();

  let page = client.get_page("123456").unwrap();
  let body = page.body.unwrap();
  let storage = body.storage.unwrap();

  assert_eq!(storage.representation, "storage");
  assert!(!storage.value.is_empty());
}

#[test]
fn test_view_format_representation() {
  let client = FakeConfluenceClient::with_sample_pages();

  let page = client.get_page("123456").unwrap();
  let body = page.body.unwrap();

  assert!(body.view.is_some());
  let view = body.view.unwrap();

  assert_eq!(view.representation, "view");
  assert!(!view.value.is_empty());
}

#[test]
fn test_error_handling_workflow() {
  let mut client = FakeConfluenceClient::with_sample_pages();

  // Test auth failure
  client.set_auth_success(false);
  assert!(client.test_auth().is_err());

  // Re-enable auth
  client.set_auth_success(true);
  assert!(client.test_auth().is_ok());

  // Test missing page
  assert!(client.get_page("nonexistent").is_err());

  // Test existing page
  assert!(client.get_page("123456").is_ok());
}

use tempfile::TempDir;

#[test]
fn test_image_download_workflow() {
  use confluence_dl::confluence::{Attachment, AttachmentLinks};
  use confluence_dl::images;

  // Create a temporary directory for the test
  let temp_dir = TempDir::new().unwrap();
  let output_path = temp_dir.path();

  // Create a fake client with sample pages
  let mut client = FakeConfluenceClient::with_sample_pages();

  // Add some attachments for the page with images (must match filenames in the
  // fixture)
  let attachments = vec![Attachment {
    id: "att1".to_string(),
    title: "architecture.png".to_string(),
    attachment_type: "attachment".to_string(),
    media_type: Some("image/png".to_string()),
    file_size: Some(1024),
    links: Some(AttachmentLinks {
      download: Some("/wiki/download/attachments/456789/architecture.png".to_string()),
    }),
  }];
  client.add_attachments("456789", attachments);

  // Get the page with images
  let page = client.get_page("456789").unwrap();
  let storage_content = page
    .body
    .as_ref()
    .and_then(|b| b.storage.as_ref())
    .map(|s| s.value.as_str())
    .unwrap();

  // Extract image references
  let image_refs = images::extract_image_references(storage_content).unwrap();
  assert!(!image_refs.is_empty(), "Should find images in the page");

  // Download images
  let filename_map = images::download_images(&client, "456789", &image_refs, output_path, "images", false).unwrap();

  // Verify images were "downloaded" (fake client creates empty files)
  assert!(!filename_map.is_empty(), "Should have downloaded images");

  // Verify files exist
  for (original_filename, local_path) in &filename_map {
    let full_path = output_path.join(local_path);
    assert!(full_path.exists(), "Image file should exist: {full_path:?}");
    println!("Downloaded {} -> {}", original_filename, local_path.display());
  }

  // Test markdown link updating
  let markdown = "![architecture](architecture.png)";
  let updated_markdown = images::update_markdown_image_links(markdown, &filename_map);

  // Verify links were updated to point to the images directory
  assert!(
    updated_markdown.contains("](images/"),
    "Links should be updated to images directory: {updated_markdown}"
  );
}

#[test]
fn test_get_child_pages_empty() {
  let client = FakeConfluenceClient::with_sample_pages();

  // Page with no children should return empty vec
  let children = client.get_child_pages("123456").unwrap();
  assert!(children.is_empty(), "Page should have no children");
}

#[test]
fn test_get_child_pages_with_children() {
  let mut client = FakeConfluenceClient::with_sample_pages();

  // Add child pages
  client.add_page_from_json("111111", fixtures::sample_child_page_1_response());
  client.add_page_from_json("222222", fixtures::sample_child_page_2_response());

  // Set up parent-child relationship
  client.add_child_pages("123456", vec!["111111".to_string(), "222222".to_string()]);

  // Get children
  let children = client.get_child_pages("123456").unwrap();
  assert_eq!(children.len(), 2, "Should have 2 children");

  // Verify child titles
  assert_eq!(children[0].title, "Child Page 1");
  assert_eq!(children[1].title, "Child Page 2");
}

#[test]
fn test_page_tree_single_page() {
  use confluence_dl::confluence::get_page_tree;

  let client = FakeConfluenceClient::with_sample_pages();

  // Build tree for page with no children
  let tree = get_page_tree(&client, "123456", None).unwrap();

  assert_eq!(tree.page.id, "123456");
  assert_eq!(tree.page.title, "Getting Started Guide");
  assert_eq!(tree.depth, 0);
  assert!(tree.children.is_empty());
}

#[test]
fn test_page_tree_with_children() {
  use confluence_dl::confluence::get_page_tree;

  let mut client = FakeConfluenceClient::with_sample_pages();

  // Set up page hierarchy: 123456 -> [111111, 222222]
  client.add_page_from_json("111111", fixtures::sample_child_page_1_response());
  client.add_page_from_json("222222", fixtures::sample_child_page_2_response());
  client.add_child_pages("123456", vec!["111111".to_string(), "222222".to_string()]);

  // Build tree
  let tree = get_page_tree(&client, "123456", None).unwrap();

  assert_eq!(tree.page.title, "Getting Started Guide");
  assert_eq!(tree.depth, 0);
  assert_eq!(tree.children.len(), 2);

  // Verify first child
  assert_eq!(tree.children[0].page.title, "Child Page 1");
  assert_eq!(tree.children[0].depth, 1);
  assert!(tree.children[0].children.is_empty());

  // Verify second child
  assert_eq!(tree.children[1].page.title, "Child Page 2");
  assert_eq!(tree.children[1].depth, 1);
  assert!(tree.children[1].children.is_empty());
}

#[test]
fn test_page_tree_with_grandchildren() {
  use confluence_dl::confluence::get_page_tree;

  let mut client = FakeConfluenceClient::with_sample_pages();

  // Set up page hierarchy: 123456 -> 111111 -> 333333
  client.add_page_from_json("111111", fixtures::sample_child_page_1_response());
  client.add_page_from_json("333333", fixtures::sample_grandchild_page_response());
  client.add_child_pages("123456", vec!["111111".to_string()]);
  client.add_child_pages("111111", vec!["333333".to_string()]);

  // Build tree with unlimited depth
  let tree = get_page_tree(&client, "123456", None).unwrap();

  assert_eq!(tree.depth, 0);
  assert_eq!(tree.children.len(), 1);

  // Verify child
  let child = &tree.children[0];
  assert_eq!(child.page.title, "Child Page 1");
  assert_eq!(child.depth, 1);
  assert_eq!(child.children.len(), 1);

  // Verify grandchild
  let grandchild = &child.children[0];
  assert_eq!(grandchild.page.title, "Grandchild Page");
  assert_eq!(grandchild.depth, 2);
  assert!(grandchild.children.is_empty());
}

#[test]
fn test_page_tree_max_depth_limit() {
  use confluence_dl::confluence::get_page_tree;

  let mut client = FakeConfluenceClient::with_sample_pages();

  // Set up page hierarchy: 123456 -> 111111 -> 333333
  client.add_page_from_json("111111", fixtures::sample_child_page_1_response());
  client.add_page_from_json("333333", fixtures::sample_grandchild_page_response());
  client.add_child_pages("123456", vec!["111111".to_string()]);
  client.add_child_pages("111111", vec!["333333".to_string()]);

  // Build tree with max_depth = 1 (should stop at children, not grandchildren)
  let tree = get_page_tree(&client, "123456", Some(1)).unwrap();

  assert_eq!(tree.depth, 0);
  assert_eq!(tree.children.len(), 1);

  // Verify child exists
  let child = &tree.children[0];
  assert_eq!(child.page.title, "Child Page 1");
  assert_eq!(child.depth, 1);

  // Grandchild should NOT be included due to depth limit
  assert!(
    child.children.is_empty(),
    "Should not fetch grandchildren when max_depth=1"
  );
}

#[test]
fn test_page_tree_depth_zero() {
  use confluence_dl::confluence::get_page_tree;

  let mut client = FakeConfluenceClient::with_sample_pages();

  // Set up children
  client.add_page_from_json("111111", fixtures::sample_child_page_1_response());
  client.add_child_pages("123456", vec!["111111".to_string()]);

  // Build tree with max_depth = 0 (should include only root page)
  let tree = get_page_tree(&client, "123456", Some(0)).unwrap();

  assert_eq!(tree.depth, 0);
  assert!(tree.children.is_empty(), "Should not fetch children when max_depth=0");
}

#[test]
fn test_page_tree_circular_reference_detection() {
  use confluence_dl::confluence::get_page_tree;

  let mut client = FakeConfluenceClient::with_sample_pages();

  // Create circular reference: 123456 -> 111111 -> 123456
  client.add_page_from_json("111111", fixtures::sample_child_page_1_response());
  client.add_child_pages("123456", vec!["111111".to_string()]);
  client.add_child_pages("111111", vec!["123456".to_string()]);

  // The function should successfully build the tree but skip the circular
  // reference (it logs a warning and continues with other children)
  let result = get_page_tree(&client, "123456", None);

  assert!(result.is_ok(), "Should handle circular reference gracefully");
  let tree = result.unwrap();

  // Root page should be present
  assert_eq!(tree.page.title, "Getting Started Guide");
  assert_eq!(tree.children.len(), 1);

  // Child should be present but without the circular reference back to parent
  let child = &tree.children[0];
  assert_eq!(child.page.title, "Child Page 1");
  assert!(child.children.is_empty(), "Circular reference should be skipped");
}

#[test]
fn test_convert_comprehensive_features_page_to_markdown() {
  use confluence_dl::markdown;

  let mut client = FakeConfluenceClient::with_sample_pages();

  // Add the comprehensive test page
  client.add_page_from_json("776655", fixtures::sample_comprehensive_features_response());

  // Fetch the page
  let page = client.get_page("776655").unwrap();

  // Extract storage content
  let storage_content = page
    .body
    .as_ref()
    .and_then(|b| b.storage.as_ref())
    .map(|s| s.value.as_str())
    .unwrap();

  // Convert to markdown
  let markdown = markdown::storage_to_markdown(storage_content).unwrap();

  assert_snapshot!(markdown);
}

#[test]
fn test_convert_meeting_notes_overview_to_markdown() {
  use confluence_dl::markdown;

  let mut client = FakeConfluenceClient::with_sample_pages();

  // Add the meeting notes overview page
  client.add_page_from_json("998877", fixtures::sample_meeting_notes_overview_response());

  // Fetch the page
  let page = client.get_page("998877").unwrap();

  // Extract storage content
  let storage_content = page
    .body
    .as_ref()
    .and_then(|b| b.storage.as_ref())
    .map(|s| s.value.as_str())
    .unwrap();

  // Convert to markdown
  let markdown = markdown::storage_to_markdown(storage_content).unwrap();

  assert_snapshot!(markdown);
}

#[test]
fn test_convert_meeting_notes_with_tasks_to_markdown() {
  use confluence_dl::markdown;

  let mut client = FakeConfluenceClient::with_sample_pages();

  // Add the meeting notes with tasks page
  client.add_page_from_json("887766", fixtures::sample_meeting_notes_with_tasks_response());

  // Fetch the page
  let page = client.get_page("887766").unwrap();

  // Extract storage content
  let storage_content = page
    .body
    .as_ref()
    .and_then(|b| b.storage.as_ref())
    .map(|s| s.value.as_str())
    .unwrap();

  // Convert to markdown
  let markdown = markdown::storage_to_markdown(storage_content).unwrap();

  assert_snapshot!(markdown);
}

#[test]
fn test_convert_complex_page_with_code_to_markdown() {
  use confluence_dl::markdown;

  let client = FakeConfluenceClient::with_sample_pages();

  // Fetch the complex page (already in sample_pages)
  let page = client.get_page("789012").unwrap();

  // Extract storage content
  let storage_content = page
    .body
    .as_ref()
    .and_then(|b| b.storage.as_ref())
    .map(|s| s.value.as_str())
    .unwrap();

  // Convert to markdown
  let markdown = markdown::storage_to_markdown(storage_content).unwrap();

  assert_snapshot!(markdown);
}

#[test]
fn test_convert_page_with_internal_links_to_markdown() {
  use confluence_dl::markdown;

  let client = FakeConfluenceClient::with_sample_pages();

  // Fetch the page with internal links
  let page = client.get_page("345678").unwrap();

  // Extract storage content
  let storage_content = page
    .body
    .as_ref()
    .and_then(|b| b.storage.as_ref())
    .map(|s| s.value.as_str())
    .unwrap();

  // Convert to markdown
  let markdown = markdown::storage_to_markdown(storage_content).unwrap();

  assert_snapshot!(markdown);
}

#[test]
fn test_end_to_end_page_fetch_and_convert() {
  use confluence_dl::markdown;

  let client = FakeConfluenceClient::with_sample_pages();

  // Test a complete workflow: fetch page -> extract content -> convert to
  // markdown
  let page_id = "123456";
  let page = client.get_page(page_id).unwrap();

  assert_eq!(page.id, page_id);
  assert_eq!(page.title, "Getting Started Guide");

  // Extract and convert
  let storage_content = page
    .body
    .as_ref()
    .and_then(|b| b.storage.as_ref())
    .map(|s| s.value.as_str())
    .expect("Page should have storage content");

  let markdown = markdown::storage_to_markdown(storage_content).unwrap();

  assert_snapshot!(markdown);
}

#[test]
fn test_markdown_conversion_handles_empty_content() {
  use confluence_dl::markdown;

  // Test that empty/minimal content doesn't crash
  let empty_storage = "";
  let result = markdown::storage_to_markdown(empty_storage);
  assert!(result.is_ok(), "Should handle empty content");

  let minimal_storage = "<p>Test</p>";
  let markdown = markdown::storage_to_markdown(minimal_storage).unwrap();
  assert!(markdown.contains("Test"), "Should contain text");
}

#[test]
fn test_markdown_conversion_preserves_structure() {
  use confluence_dl::markdown;

  let mut client = FakeConfluenceClient::with_sample_pages();

  // Add the comprehensive page
  client.add_page_from_json("776655", fixtures::sample_comprehensive_features_response());
  let page = client.get_page("776655").unwrap();

  let storage_content = page
    .body
    .as_ref()
    .and_then(|b| b.storage.as_ref())
    .map(|s| s.value.as_str())
    .unwrap();

  let markdown = markdown::storage_to_markdown(storage_content).unwrap();

  assert_snapshot!(markdown);
}
