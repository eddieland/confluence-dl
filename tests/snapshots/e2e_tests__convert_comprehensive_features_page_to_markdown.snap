---
source: tests/e2e_tests.rs
assertion_line: 511
expression: markdown
---
# Comprehensive Test Page

## Basic Formatting

This page contains **bold**, _italic_, and _underlined_ text.

It also has `inline code` and [external links](https://example.com).

## Lists

- Unordered item 1
- Unordered item 2
  - Nested item

1. Ordered item 1
2. Ordered item 2

## Emoticons and Icons

✅ Success indicator

❌ Failure indicator

## Internal Links

See [[Getting Started Guide]] for more information.

## User Mentions

Assigned to @user:6320c26429083bbe8cc369b0 for review.

## Task Lists

- [x] Complete this task
- [ ] Pending task

## Tables

| Header 1 | Header 2 |
| -------- | -------- |
| Cell 1   | Cell 2   |

## Code Blocks

```javascript
function greet(name) { console.log(name); }
```

## Images

![image]()

## Placeholders

## Time Elements

Meeting date: 2025-10-07
